ta/crypto_perf/ta_crypto_perf.c

```diff
@@ -465,8 +467,479 @@ TEE_Result cmd_hash_prepare_op(uint32_t param_types, TEE_Param params[4])
+			EMSG("ECDSA error mode");
+			res = TEE_ERROR_BAD_PARAMETERS;
+		}
+	}
+
+	return res;
+}
+
+static uint32_t get_keypair_type(uint32_t value)
+{
+	switch (value) {
+	case DH:
+		return TEE_TYPE_DH_KEYPAIR;
+	case RSA:
+		return TEE_TYPE_RSA_KEYPAIR;
+	case ECDSA:
+		return TEE_TYPE_ECDSA_KEYPAIR;
+	case ECDH:
+		return TEE_TYPE_ECDH_KEYPAIR;
+	case X25519:
+		return TEE_TYPE_X25519_KEYPAIR;
+	default:
+		EMSG("The algo[%u] is not valid", algo);
+	}
+
+	return TEE_TYPE_ILLEGAL_VALUE;
+}
+
+TEE_Result cmd_asym_process_keypair(uint32_t param_types,
+			       TEE_Param params[TEE_NUM_PARAMS])
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	TEE_Attribute *attrs = NULL;
+	uint32_t attr_count = 0;
+	int width_bits = 0;
+	int n = 0;
+	uint32_t exp_param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+						   TEE_PARAM_TYPE_MEMREF_INOUT,
+						   TEE_PARAM_TYPE_NONE,
+						   TEE_PARAM_TYPE_NONE);
+
+	if (param_types != exp_param_types)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = unpack_attrs(params[1].memref.buffer, params[1].memref.size,
+			   &attrs, &attr_count);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	width_bits = params[0].value.a;
+	n = params[0].value.b;
+
+	while (n--) {
+		res = TEE_GenerateKey(crypto_obj, width_bits, attrs, attr_count);
+		CHECK(res, "TEE_GenerateKey", goto out;);
+		TEE_ResetTransientObject(crypto_obj);
+	}
+
+out:
+	TEE_Free(attrs);
+	return res;
+}
+
+TEE_Result cmd_asym_process_rsa_ecc(uint32_t param_types,
+				    TEE_Param params[TEE_NUM_PARAMS])
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	int n = 0;
+	uint32_t mode = 0;
+	uint32_t exp_param_types = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+						   TEE_PARAM_TYPE_VALUE_INPUT,
+						   TEE_PARAM_TYPE_MEMREF_INPUT,
+						   TEE_PARAM_TYPE_MEMREF_OUTPUT);
+	TEE_Attribute *attrs = NULL;
+	uint32_t attr_count = 0;
+	TEE_Result (*do_asym)(TEE_OperationHandle, const TEE_Attribute *,
+			      uint32_t, const void *, uint32_t, void *,
+			      uint32_t *) = NULL;
+
+	if (param_types != exp_param_types)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = unpack_attrs(params[0].memref.buffer, params[0].memref.size,
+			   &attrs, &attr_count);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	n = params[1].value.a;
+	mode = params[1].value.b;
+
+	if (mode == MODE_ENCRYPT)
+		do_asym = TEE_AsymmetricEncrypt;
+	else if (mode == MODE_DECRYPT)
+		do_asym = TEE_AsymmetricDecrypt;
+	else if (mode == MODE_SIGN)
+		do_asym = TEE_AsymmetricSignDigest;
+
+	if (mode == MODE_VERIFY) {
+		while (n--) {
+			res = TEE_AsymmetricVerifyDigest(crypto_op, attrs,
+				attr_count, params[2].memref.buffer,
+				params[2].memref.size, params[3].memref.buffer,
+				params[3].memref.size);
+			CHECK(res, "TEE_AsymmetricVerifyDigest", goto out;);
+		}
+	} else {
+		while (n--) {
+			res = do_asym(crypto_op, attrs, attr_count,
+				params[2].memref.buffer, params[2].memref.size,
+				params[3].memref.buffer, &params[3].memref.size);
+
+			CHECK(res, "TEE_AsymmetricEncrypt", goto out;);
```

Comment:
for consistency and simplicity: `CHECK(res, "TEE proccessing failed", goto out;);`